zone/parcels.cpp

```diff
@@ -0,0 +1,755 @@
+			auto result = ParcelsRepository::InsertOne(database, parcel_out);
+			if (!result.id) {
+				LogError(
+					"Failed to add parcel to database.  From {} to {} item {} quantity {}",
+					parcel_out.from_name,
+					parcel_out.to_name,
+					parcel_out.item_id,
+					parcel_out.quantity
+				);
+				Message(
+					Chat::Yellow,
+					"Unable to save parcel to the database. Please see an administrator."
+				);
+				return;
+			}
+
+			MessageString(
+				Chat::Yellow,
+				PARCEL_DELIVERY,
+				merchant->GetCleanName(),
+				money.c_str(),
+				send_to_client.at(0).character_name.c_str()
+			);
+
+			if (player_event_logs.IsEventEnabled(PlayerEvent::PARCEL_SEND)) {
+				PlayerEvent::ParcelSend e{};
+				e.from_player_name = parcel_out.from_name;
+				e.to_player_name   = parcel_out.to_name;
+				e.item_id          = parcel_out.item_id;
+				e.quantity         = parcel_out.quantity;
+				e.sent_date        = parcel_out.sent_date;
+
+				RecordPlayerEventLog(PlayerEvent::PARCEL_SEND, e);
+			}
+
+			m_parcel_platinum = 0;
+			m_parcel_gold     = 0;
+			m_parcel_silver   = 0;
+			m_parcel_copper   = 0;
+			auto outapp = new EQApplicationPacket(OP_FinishTrade);
+			FastQueuePacket(&outapp);
+
+			Parcel_Struct ps{};
+			ps.item_slot = parcel_out.slot_id;
+			strn0cpy(ps.send_to, parcel_out.to_name.c_str(), sizeof(ps.send_to));
+
+			SendParcelDeliveryToWorld(ps);
+
+			break;
+		}
+	}
+
+	SendParcelAck();
+	SendParcelIconStatus();
+	SetParcelEnabled(false);
+	GetParcelTimer()->Enable();
+}
+
+void Client::SendParcelAck()
+{
+	auto outapp = new EQApplicationPacket(OP_FinishTrade);
+	FastQueuePacket(&outapp);
+
+	outapp = new EQApplicationPacket(OP_ShopSendParcel, sizeof(Parcel_Struct));
+	auto data = (Parcel_Struct *) outapp->pBuffer;
+	data->item_slot = 0xffffffff;
+	data->quantity  = 0xffffffff;
+	FastQueuePacket(&outapp);
+}
+
+void Client::SendParcelRetrieveAck()
+{
+	auto outapp = new EQApplicationPacket(OP_ShopRetrieveParcel);
+	FastQueuePacket(&outapp);
+}
+
+void Client::SendParcelDeliveryToWorld(Parcel_Struct parcel)
+{
+	auto out  = new ServerPacket(ServerOP_ParcelDelivery, sizeof(Parcel_Struct));
+	auto data = (Parcel_Struct *) out->pBuffer;
+
+	data->item_slot = parcel.item_slot;
+	strn0cpy(data->send_to, parcel.send_to, sizeof(data->send_to));
+
+	worldserver.SendPacket(out);
+}
+
+void Client::DoParcelRetrieve(ParcelRetrieve_Struct parcel_in)
+{
+	auto merchant = entity_list.GetNPCByID(parcel_in.merchant_entity_id);
+	if (!merchant) {
+		SendParcelRetrieveAck();
+		return;
+	}
+
+	auto p = m_parcels.find(parcel_in.parcel_slot_id);
+	if (p != m_parcels.end()) {
+		uint32 item_id       = parcel_in.parcel_item_id;
+		uint32 item_quantity = p->second.quantity;
+		if (!item_id || !item_quantity) {
+			LogError(
+				"Attempt to retrieve parcel with erroneous item id or quantity for client character id {}.",
+				CharacterID()
+			);
+			SendParcelRetrieveAck();
+			return;
+		}
+
+		auto inst = database.CreateItem(item_id, item_quantity);
```

Comment:
`inst` needs to be deleted. Even though `TryStacking` and `PutItemInInventory` are called with this instance below, those functions do not take ownership and just clones it, so this still needs to be freed after

Reply:
As noted above, converted to unique_ptr.